### PR TITLE
Alex/fixforqueryrefresh

### LINKED ISSUE #8891: Edit Data grid never displays if Run is clicked many times
- Azure Data Studio Version: `1.14.1`

Steps to Reproduce:

1. Open Edit Data on a table
2. Show the SQL Pane
3. Click the Run button over-and-over numerous times (e.g. ~5-10 times) without waiting for previous Run to complete 

Expected: Once done clicking the Run button the result grid displays respecting the current SQL query

Actual: The grid never displays.  The Run button is disabled and Stop is enabled.  Clicking Stop does nothing and editor must be closed and reopened to be usable again.

image

### PR DESCRIPTION
**Pull request by smartguest**
if constant refreshes occur when in the loadDataFunction, the dataSet in it becomes undefined in the middle of the process (even with a check to ensure that the dataSet exists outside) and throws an error, resulting in the grid not loading any data whatsoever and only having null in it. To handle this and a similar issue in refreshGrid, try catch statements are used to catch the error when it occurs and to not alter the grid at all when it happens, only displaying an error message. This preserves the grid and does not wipe it out when zero data is loaded from the process.

This PR fixes #8891

### CODE REVIEW

File: src/sql/workbench/contrib/editData/browser/editDataGridPanel.ts [hunk 1]
```diff
@@ -214,32 +215,44 @@ export class EditDataGridPanel extends GridParentComponent {
...
-						return p;
-					}, {}));
-				}
 
-				return gridData;
-			});
+						return gridData;
+					}
+					catch {
+						throw new Error('Failed to load table data');
```
Comment (kburtram) [inline src/sql/workbench/contrib/editData/browser/editDataGridPanel.ts]: why do we catch here if we're just rethrow?

Comment (smartguest) [reply src/sql/workbench/contrib/editData/browser/editDataGridPanel.ts]: This is meant to show a message to the user that the table data failed to load, probably best to raise an error instead.

File: src/sql/workbench/contrib/editData/browser/editDataGridPanel.ts [hunk 2]
```diff
@@ -218,32 +219,44 @@ export class EditDataGridPanel extends GridParentComponent {
...
-					}, {}));
-				}
 
-				return gridData;
-			});
+						return gridData;
+					}
+					catch (e) {
+						//throw exception so that grid will not be updated with bad data.
+						throw e;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/editData/browser/editDataGridPanel.ts:252]: Why bother catching if you're just rethrowing?

File: src/sql/workbench/contrib/editData/browser/editDataGridPanel.ts [hunk 3]
```diff
@@ -435,16 +448,20 @@ export class EditDataGridPanel extends GridParentComponent {
...
-				if (this.oldDataRows !== this.placeHolderDataSets[0].dataRows) {
-					this.detectChange();
-					this.oldDataRows = this.placeHolderDataSets[0].dataRows;
+					if (this.placeHolderDataSets[0].dataRows && this.oldDataRows !== this.placeHolderDataSets[0].dataRows) {
+						this.detectChange();
+						this.oldDataRows = this.placeHolderDataSets[0].dataRows;
+					}
+				}
+				catch {
+					this.notificationService.error(nls.localize('refreshTableError', 'Too many refreshes, edit session corrupt: Please close and reload table'));
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/editData/browser/editDataGridPanel.ts:464]: Do they really have to close and reopen the table? Your description seems to imply that the table will be fine and so shouldn't they just be able to refresh again at that point?

File: src/sql/workbench/contrib/editData/browser/editDataGridPanel.ts (hunk 2 above)
Comment (smartguest) [reply src/sql/workbench/contrib/editData/browser/editDataGridPanel.ts:252]: Nevermind, it turns out I added a redundant outer try catch condition, I just found out you can directly reject the promise within the inner try catch condition.

File: src/sql/workbench/contrib/editData/browser/editDataGridPanel.ts (hunk 3 above)
Comment (smartguest) [reply src/sql/workbench/contrib/editData/browser/editDataGridPanel.ts:464]: The issue is that when you refresh too many times, an error occurs in vscode-jsonrpc module that states an edit session has not been initialized.

**Comment by smartguest** [comment]
The try catch block is now fixed for refreshGrid, if it catches an error resulting from the data set being null, the promise to refresh the grid is rejected.  For loadDataFunction, the last known good grid data is returned in case the data set to load from is null

**Pull request merged**